### DISCUSSION TITLE
Can I use the same port for multiple input harbors, like audio & studio cam?

### DISCUSSION THREAD

**Discussion by Moonbase59** [savonet/liquidsoap#3370] [Q&A] [answered]
I plan to not only let DJs/Show hosts stream in _audio_, but also their _studio cam_ (or maybe a weather cam), if available. This will be a nice extra for my video stream.

Now I wonder if it is "officially" okay to use the _same input port_ for _multiple input harbors_ (i.e., one audio, one cam), because I don’t want to rip open my network for too many ports. **So, for example, a studio would stream in their show via Icecast protocol on 8005, or Shoutcast protocol on 8005/8006, could they _also_ stream in their studio cam (say, via an _ffmpeg_ script) on port 8005 (using another mount name)?**

Interestingly, this _seems_ to work, but **I’d like to get the dev’s official statement if this is supported and _should_ really be working.** Here’s part of the code I tested:

```ruby
... (other code) ...

# Live Broadcasting

# Radio
live = input.harbor("/live", id = "input_streamer", port = 8005, auth = dj_auth, icy = true, icy_metadata_charset = "UTF-8", metadata_charset = "UTF-8", on_connect = live_connected, on_disconnect = live_disconnected, buffer = 5.00, max = 10.00)

# Studio Cam
cam = input.harbor("/cam", id = "input_studiocam", port = 8005, auth = vj_auth, icy = true, icy_metadata_charset = "UTF-8", metadata_charset = "UTF-8", on_connect = cam_connected, on_disconnect = cam_disconnected, buffer = 5.00, max = 10.00)

... (other code) ...

# Annotate a (default) duration of 15s, in case images are used here.
# For videos, the duration will be overwritten by the video’s length.
v = mksafe(playlist(
  prefix="annotate:duration=15.0,liq_video_fade_in=0.2,liq_video_fade_out=0.2:",
  reload_mode="watch",
  video_playlist))

# if cam input comes in, switch immediately
v = fallback(track_sensitive=false, [cam, v])

... (other code) ...
```

Super-simple _ffmpeg_ script for testing:
```bash
#!/bin/bash

# restream Cam1 into station 1’s video stream

# camera 1 (can't keep up more than 5-7 fps; it's a weather cam)
ffmpeg \
  -r 15 -i "http://cam1:8081" \
  -f lavfi -i anullsrc \
  -c:v libx264 -vf format=yuv420p -r 15 -preset ultrafast -tune zerolatency -g 30 \
    -maxrate 2000k -bufsize 4000k \
  -c:a aac -b:a 128k -ar 44100 -ac 2                              \
  -f flv \
  icecast://source:hackme@example.com:8005/cam


# a local webcam (Logitech HD Pro Webcam C920)
# using -input_format h264 is FAST but only works if the cam can output H.264
# check: ffplay -f v4l2 -list_formats all /dev/video2
# the cam actually can output 24 or 30 fps, NOT 25 fps
# using MJPEG @ 15fps seems a good compromise that most cams can handle
ffmpeg \
  -re \
  -f v4l2 -input_format mjpeg -video_size 1280x720 -framerate 15 -i /dev/video2 \
  -f lavfi -i anullsrc \
  -c:v libx264 -vf format=yuv420p -preset ultrafast -tune zerolatency -g 30 \
    -maxrate 3000k -bufsize 6000k \
  -c:a aac -b:a 128k -ar 44100 -ac 2                              \
  -f flv \
  icecast://source:hackme@example.com:8005/cam
```

---

Sample using a (bad) Raspberry Pi Zero W cam over WiFi, while at the same time streaming via IDJC to the same port, using above code:

![vlcsnap-2023-08-30-15h41m42s373](https://github.com/savonet/liquidsoap/assets/3706922/2b875851-d105-41d6-9783-94c3a6d7c50c)

**Comment by toots** [comment, marked as answer]
As far as I remember, harbot inputs register HTTP endpoints on the underlying HTTP server much like any other handler so having multiple on the same port shouldn't be a problem.

The one thing we do not do at the moment is multiple _output.harbor_ on the same port. This would need a refactorization of the code and, frankly, `output.harbor` is either in need of TLC or to be officially deprecated.

**Comment by Moonbase59** [comment]
That _is_ good news, thank you!

…and please _don’t_ deprecate `output.harbor`! It’s surely not meant as an Icecast replacement, but there _are_ uses, be it only some quick-n-dirty testing, or for some local restreaming…

"TLC" surely means "Thunder, Lightning, and a Cat", right? Or did you mean "tender loving care" after all? :rofl:

**Reply by toots** [reply]
Some lovin'!
